src/Core/Grid/Data/Factory/AttachmentGridDataFactoryDecorator.php

```diff
@@ -0,0 +1,160 @@
+ */
+
+namespace PrestaShop\PrestaShop\Core\Grid\Data\Factory;
+
+use Doctrine\DBAL\Connection;
+use PrestaShop\PrestaShop\Core\Grid\Data\GridData;
+use PrestaShop\PrestaShop\Core\Grid\Record\RecordCollection;
+use PrestaShop\PrestaShop\Core\Grid\Record\RecordCollectionInterface;
+use PrestaShop\PrestaShop\Core\Grid\Search\SearchCriteriaInterface;
+use PrestaShop\PrestaShop\Core\Util\File\FileSizeConverter;
+use PrestaShopBundle\Translation\TranslatorAwareTrait;
+
+/**
+ * Decorates attachment grid data factory
+ */
+final class AttachmentGridDataFactoryDecorator implements GridDataFactoryInterface
+{
+    use TranslatorAwareTrait;
+
+    /**
+     * @var GridDataFactoryInterface
+     */
+    private $attachmentDoctrineGridDataFactory;
+
+    /**
+     * @var int
+     */
+    private $employeeIdLang;
+
+    /**
+     * @var Connection
+     */
+    private $connection;
+
+    /**
+     * @var string
+     */
+    private $dbPrefix;
+
+    /**
+     * @var FileSizeConverter
+     */
+    private $fileSizeConverter;
+
+    /**
+     * @param GridDataFactoryInterface $attachmentDoctrineGridDataFactory
+     * @param int $employeeIdLang
+     * @param Connection $connection
+     * @param string $dbPrefix
+     * @param FileSizeConverter $fileSizeConverter
+     */
+    public function __construct(
+        GridDataFactoryInterface $attachmentDoctrineGridDataFactory,
+        int $employeeIdLang,
+        Connection $connection,
+        string $dbPrefix,
+        FileSizeConverter $fileSizeConverter
+    ) {
+        $this->attachmentDoctrineGridDataFactory = $attachmentDoctrineGridDataFactory;
+        $this->employeeIdLang = $employeeIdLang;
+        $this->connection = $connection;
+        $this->dbPrefix = $dbPrefix;
+        $this->fileSizeConverter = $fileSizeConverter;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getData(SearchCriteriaInterface $searchCriteria)
+    {
+        $attachmentData = $this->attachmentDoctrineGridDataFactory->getData($searchCriteria);
+
+        $attachmentRecords = $this->applyModifications($attachmentData->getRecords());
+
+        return new GridData(
+            $attachmentRecords,
+            $attachmentData->getRecordsTotal(),
+            $attachmentData->getQuery()
+        );
+    }
+
+    /**
+     * @param RecordCollectionInterface $attachments
+     *
+     * @return RecordCollection
+     */
+    private function applyModifications(RecordCollectionInterface $attachments)
+    {
+        $modifiedAttachments = [];
+
+        foreach ($attachments as $attachment) {
+            if ((int) $attachment['products'] > 0) {
+                $productNamesArray = $this->getProductNames($attachment['id_attachment']);
+                $productNames = $productNamesArray['product_names'] ?? '';
+                $attachment['dynamic_message'] = $this->trans(
+                    'This file is associated with the following products, do you really want to  delete it?',
```

Comment:
It seems that there is a useless space in the original string, but I'm not fond of letting it as-is because it is a mistake. In order to bypass the issue, I would suggest writing the following new wording _This file is associated with the following %products%. Are you sure you want to delete it?_, what do you think?

Reply:
Sound good to me